homeassistant/helpers/intent.py

```diff
@@ -117,6 +124,72 @@ def __repr__(self):
         return '<{} - {}>'.format(self.__class__.__name__, self.intent_type)
 
 
+def fuzzymatch(name, entities):
+    """Fuzzy matching function."""
+    matches = []
+    pattern = '.*?'.join(name)
+    regex = re.compile(pattern)
+    for entity in entities:
+        match = regex.search(entity)
+        if match:
+            matches.append((len(match.group()), match.start(), entity))
+    return [x for _, _, x in sorted(matches)]
+
+
+class ServiceIntentHandler(IntentHandler):
+    """Service Intent handler registration.
+
+    Service specific intent handler that calls a service by name/entity_id.
+    """
+
+    slot_schema = {
+        'name': cv.string,
+    }
+
+    def __init__(self, intent_type, domain, service, speech):
+        """Create Service Intent Handler."""
+        self.intent_type = intent_type
+        self.domain = domain
+        self.service = service
+        self.speech = speech
+
+    @asyncio.coroutine
+    def async_handle(self, intent_obj):
+        """Handle the hass intent."""
+        hass = intent_obj.hass
+        slots = self.async_validate_slots(intent_obj.slots)
+        response = intent_obj.create_response()
+
+        name = slots['name']['value']
+        entities = {state.entity_id: state.name for state
+                    in hass.states.async_all()}
+        entity_name = name.replace(' ', '_').lower()
+        entity_name = entity_name.replace('the_', '')
+
+        matches = fuzzymatch(entity_name, entities)
+        entity_id = next((x for x in matches if self.domain in x), None)
```

Comment:
What's with this check? Isn't domain for now always `homeassistant` ?

Reply:
Well my plan is to add things like covers and media players, so an Intent like HassOpenCover declares it's domain we then match against cover.garage_door and not switch.garage_door or sensor.garage_door.

Reply:
I think that it's:
- out of scope for the current implementation (since it only works with `homeassistant`
- Filter should be applied when generating the dictionary of potential entities to match instead of filtering it after matching.
- A "limit matching to domain" should be an optional boolean passed into the constructor

Reply:
Why would we add another passed variable instead of just 'if domain not Homeassistant'? And I don't think it's out of scope since the whole intention of this is to lay some groundwork for adding other intents. Otherwise all of this has just been rearranging things with zero added value beyond cleanup.

Reply:
Can you 100% guarantee that you can always derive the filter from the service domain? Because if you can't, an extra variable is necessary.

A cleanup so that you can start building on top is perfectly fine to be in a single pr. If you're adding functionality you need to use it and you need to test it. Neither is happening.

Reply:
The best filter would be to filter by entites that support the given domain.service

Reply:
Correct. However, one thing is that Home Assistant core should not be aware of the different components. So if we would want something like this, we would have to make it part of service registration, and then the service registry needs to know about which attribute should map to the entity… it's a bit too much. So I think that passing the right filter at the moment we define the intent is great, that way the logic remains inside the components. (light intent will specify light filter, etc)